### PR TITLE
[auto-materialize][3/n] Recover previous materialization results and merge them

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Builds on: https://github.com/dagster-io/dagster/pull/17075

This PR finally makes some logical changes (which should not impact any existing rules). In essence, whenever an asset enters the "skipped" state, its materialize rules will continually fire on each tick until it is materialized (either by the Asset Daemon or manually).

In order to make this efficient, we allow materialize rules to simply dig into their history and return their prior evaluation results, so no new computation needs to happen. For skip rules, they are allowed to choose which candidates they actually need to re-evaluate for. For all existing skip rules, the choice is simple -- only re-evaluate brand-new candidates, or old candidates whose parents have / will update.

For example, if you've decided to skip some partition of an asset because its parents are missing, you don't need to do any work to determine that its parents are still missing if none of its parents have been updated since the last tick.

In practice, this does not meaningfully impact runtime, even if there are large sets of assets in the skipped state (adds ~1 second per tick for ~3000 asset partitions in the skipped state).

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 1]
```diff
@@ -215,6 +257,53 @@ def description(self) -> str:
...
 
+    def merge_with_past_results(
+        self,
+        context: RuleEvaluationContext,
+        asset_partitions_by_evaluation_data: Mapping[
+            Optional[AutoMaterializeRuleEvaluationData], AbstractSet[AssetKeyPartitionKey]
+        ],
+        evaluated_candidates: Optional[AbstractSet[AssetKeyPartitionKey]] = None,
+    ) -> RuleEvaluationResults:
+        """Merge the new results of an evaluation with any previous results which are still valid,
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:268]: What exactly does "merge" mean? And what about "still valid" mean? I'm guessing that "merge" means taking the union of the asset partitions?

**Comment by sryza** [comment]
> In practice, this does not meaningfully impact runtime, even if there are large sets of assets in the skipped state (adds ~1 second per tick for ~3000 assets in the skipped state).

How many partitions per asset?

**Comment by OwenKephart** [comment]
> > In practice, this does not meaningfully impact runtime, even if there are large sets of assets in the skipped state (adds ~1 second per tick for ~3000 assets in the skipped state).
> 
> How many partitions per asset?

Whoops, I meant "asset partitions" there -- edited.

**Comment by sryza** [comment]
Oh or is the idea that we wouldn't expect it to scale with the number of partitoins?

**Comment by sryza** [comment]
> Whoops, I meant "asset partitions" there -- edited.

Ah gotcha. How would you expect this to scale? We have some users with like a million total asset partitions, right?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:268]: maybe:

```
Combine the new results of an evaluation with any results from the previous tick which are still relevant.
```

or

```
Combine the new results of an evaluation with any results from the previous tick for asset partitions which are still unhandled.
```

**Review by sryza — COMMENTED** [review]
Thanks for breaking these PRs up like this. I'm getting much closer to understanding. And thanks for bearing with me on this.

I'm trying to reason about this from first principles to:
- Help me understand how it should work so I can have useful opinions
- Build confidence that the way auto-materialize rules work and fit together is legible to future readers of this code and users trying to understand how AMPs function

Here's my attempt:

To get to a performant auto-materialize rule, you kind of basically need to think about two things:
- What the rule thought on its last tick
- What has changed since that tick

### Materialize rule example

Taking for example `MaterializeOnParentUpdatedRule`. This rule should return true for an asset partition if that asset partition has parent asset partitions that have been updated since the last time the asset partition was asset_materialization_planned, requested, or discarded.

So assuming you know whether an asset partition satisfied this rule on its last tick, it should return true if: ((it satisfied this condition on its last tick and wasn't asset_materialization_planned, requested, or discarded since then) or (one of its parents update since its last tick, and it wasn't asset_materialization_planned, requested, or discarded since that update)).

### Skip rule example

Taking for example `SkipOnParentMissingRule`. This rule should return true for an asset partition if that asset partition has any missing parents that aren't slated to be materialized as part of the current tick.

So assuming we know whether an asset partition satisfied this rule on its last tick, it should return true if: it satisfied this rule on its last tick or all of its parents that were missing on the last tick (have been materialized since then or are slated to be materialized as part of the current tick).

If we add support for wiping asset partitions, then we'd return false if any of its parents had been wiped since the last tick and hadn't been materialized after that.

### Thoughts

So I'm basically trying to understand how different the current PR is from the logic expressed above.

A couple of the differences I'm noticing:
- If I understand correctly, there's some degree of preprocessing that we're doing, with `skipped_asset_graph_subset`, before passing candidates into the skip rules. Is this necessary? What do we lose by expressing the skip rules like above? I worry that with the current approach there's a hidden set of assumptions we're making.
- Is `merge_with_past_results(x)` exactly equivalent to "(it satisfied this condition on its last tick and wasn't asset_materialization_planned, requested, or discarded since then) or (x)"? If so, can we name and docstring this method in a precise way that gets this across? Methods that do a composite of a few different things should often have long names to make clear all the things they're doing.

General thought based on the last point: if we can avoid words like "valid", "handled", and "relevant" that aren't included in AutoMaterializePolicy docs, I think it will make the code much easier to work with.  Of course in some cases we need to come up with new words that represent composites that show up together repeatedly, but I think that in those cases we should find a central place to write down a precise definition of that word.

**Comment by OwenKephart** [comment]
@sryza 

> If I understand correctly, there's some degree of preprocessing that we're doing, with skipped_asset_graph_subset, before passing candidates into the skip rules. Is this necessary? What do we lose by expressing the skip rules like above? I worry that with the current approach there's a hidden set of assumptions we're making.

This is a good callout -- along these lines, I've changed this to just purely be an asset graph subset that tracks the precise set of asset partitions which were skipped last tick. That simplifies a bit of code, as we don't need to manually update the subset by removing materialized / discarded partitions (we just bundle up the skipped partitions and make a subset out of that)

> Is merge_with_past_results(x) exactly equivalent to "(it satisfied this condition on its last tick and wasn't asset_materialization_planned, requested, or discarded since then) or (x)"? If so, can we name and docstring this method in a precise way that gets this across? Methods that do a composite of a few different things should often have long names to make clear all the things they're doing.

That's basically it yep. I've refactored things here, with `add_evaluation_data_from_previous_tick` now serving as the entrypoint for this behavior. It is much more flexible than previous iterations, as you can just pass in an arbitrary function to tell the framework which asset partitions should still be considered from the previous tick. In all cases so far, that function is pretty much the same, but this makes it more explicit what's going on.

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py [hunk 2]
```diff
@@ -353,9 +352,12 @@ def evaluate_asset(
                 to_materialize.update(asset_partitions)
             self._logger.debug("Done evaluating materialize rule")
 
-        # These should be conditions, but aren't currently, so we just manually strip out things
+        # these should be conditions, but aren't currently, so we just manually strip out things
         # from our materialization set
         for candidate in list(to_materialize):
+            # ignore any partitions that were skipped on the last tick
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:358]: What if a candidate that was skipped on the last tick would no longer be skipped on this tick? E.g. if we have a skip rule that avoids materializing anything before 9 am, and the last tick is before 9 am and the next tick is after 9 am?

**Review by sryza — COMMENTED** [review]
@OwenKephart thanks for the clarifications. I dug a little deeper and left some comments on more granular pieces.

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 3]
```diff
@@ -215,6 +283,31 @@ def description(self) -> str:
...
+            asset_partitions_by_evaluation_data: A mapping from evaluation data to the set of asset
+                partitions that the rule applies to.
+            should_include: A function that returns whether a given asset partition from the
+                previous tick should be included in the results of this tick.
+        """
+        asset_partitions_by_evaluation_data = defaultdict(set, asset_partitions_by_evaluation_data)
+        for evaluation_data, asset_partitions in context.get_previous_tick_results(self):
+            for ap in asset_partitions:
+                if should_include(ap):
+                    asset_partitions_by_evaluation_data[evaluation_data].add(ap)
```
[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:307]: That would not be kosher, and currently the way we're avoiding this is for the caller of this function to pass in a `should_include` which filters out the set of asset partitions that it calculated new values for on this tick.

However, this is a bit of a landmine, so I'll update this function to explicitly filter for potential duplicates.

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:358]: that comment is confusing and I'll update it, but that line is just saying "do not check things that we skipped last tick for the weird 'not really discard rules' of 'currently being backfilled' and 'has invalid parents', because if they were valid last tick they're still valid now." basically just a performance boost (as it can be a bit expensive to check that all of the parents of a bunch of skipped asset partitions are still valid).

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:358]: Still not 100% grokking - what if a backfill starts in between the last tick and the current tick? We'd want to remove it from `to_materialize` in that case, right?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 4]
```diff
@@ -133,6 +134,32 @@ class RuleEvaluationContext(NamedTuple):
     def asset_graph(self) -> AssetGraph:
         return self.instance_queryer.asset_graph
 
+    @property
+    def new_candidates(self) -> AbstractSet[AssetKeyPartitionKey]:
+        """The set of candidates on this tick which were not candidates on the previous tick."""
+        return {c for c in self.candidates if c not in self.cursor.skipped_on_last_tick_subset}
```
[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:140]: Yeah this is a docstring / naming weirdness. Renamed this to be more explicit

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 5]
```diff
@@ -444,8 +542,14 @@ def description(self) -> str:
         return "waiting on upstream data to be up to date"
 
     def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationResults:
-        asset_partitions_by_waiting_on_asset_keys = defaultdict(set)
-        for candidate in context.candidates:
+        asset_partitions_by_evaluation_data = defaultdict(set)
+
+        # only need to evaluate net-new candidates and candidates whose parents have changed
+        candidates_to_evaluate = (
+            context.get_candidates_not_skipped_on_previous_tick()
```
[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:549]: Those results from the previous tick won't be accurate if either:

a) the candidate wasn't skipped on the previous tick (meaning no skip evaluation data from the previous tick is relevant)

or

b) one or more of the candidate's parents have been updated since the previous tick (as this can invalidate the previously-calculated result)

[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:549]: To clarify, if I understand correctly, the cursor trscks things:
1. For this individual skip rule, which asset partitions were skipped on the previous tick
2. Across all rules, which assets partitions were skipped on the previous tick

I wondering why we're using (2) instead of (1). My reasons for preferring (1):
- Feels easier to reason about. I only need to think about what happened in this rule evaluation, not what happened in all the other rule evaluations.
- Maybe we don't need to track the global skip set on the cursor at all?

[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:553]: gah you're very right

[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:553]: specifically, we do actually need some information that's global across rules, but we do not need to store this information in a separate object on the cursor as the asset evaluations contain all the information we need.

the reason we need this "rule-spanning" information is that we can only re-use data for an asset if it wasn't requested or discarded on the previous tick, and determining if a partition was requested or discarded requires knowledge of what all of the rules evaluated to.

[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: > the reason we need this "rule-spanning" information is that we can only re-use data for an asset if it wasn't requested or discarded on the previous tick

why exactly is that?

[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: I can imagine cases where some rules don't want this behavior, but for materialize_on_missing / materialize_on_parent_updated, this is the case. In particular, imagine the following sequence, for `A -> B`, where B just has "materialize_on_missing" and "skip_on_parent_missing", and A has no policy:

On tick 1, B notices that it is missing, but also that A is missing, meaning both rules fire, and no materialization is requested.

Things remain in this state (both rules firing, no asset requested) until A is manually executed. During this time, the materialize_on_missing rule is continually firing because we remember its value from the previous tick.

On the next tick, we re-evaluate the skip rule, find that it is now false, and request to materialize B (because the materialize rule is still firing).

On the next tick after that, we now need to know that we no longer should fire that materialize rule. The reason that we no longer want to fire it is that we requested it on the previous tick. In order to determine that it was requested, we need to either keep track of that in the cursor (previous solution), or re-derive its status from the asset evaluation (current solution).

[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: Gotcha. I see why, for `materialize_on_missing` and `materialize_on_parent_updated`, we need to check whether the asset partition was requested or discarded on a previous tick.

Do we need to do check that in the skip rules though?

[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: Side question: in `materialize_on_missing` and `materialize_on_parent_updated`, how _do_ we check whether the asset partition was requested or discarded on a previous tick?

File: python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/scenarios/auto_materialize_policy_scenarios.py [hunk 6]
```diff
@@ -236,6 +236,7 @@
                 )
             ],
             expected_evaluations=[
+                AssetEvaluationSpec.empty("daily"),
```
Comment (OwenKephart) [inline python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/scenarios/auto_materialize_policy_scenarios.py:239]: all of these changes are due to the fact that we're now writing all asset evaluations (empty or otherwise) in cases where the new evaluation is different from the previous one.

the end result is that in most cases, we'll get an empty evaluation written the tick after an asset is requested, as there will no longer be a reason to materialize

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 7]
```diff
@@ -119,7 +122,8 @@ class AutoMaterializeRuleEvaluation(NamedTuple):
     evaluation_data: Optional[AutoMaterializeRuleEvaluationData]
 
 
-class RuleEvaluationContext(NamedTuple):
+@dataclass
+class RuleEvaluationContext:
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:126]: turned this into a dataclass so that I could use cached_property on it.

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:126]: We should in general use `frozen=True` for dataclasses I believe

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py]: Just circling back on this one, what's going on here is just a jank version of the candidate filtering that happens in other "actual" rules. You're correct that for the backfill case, we'd want to have different candidate filtering behavior vs. the missing parent partition case.

I updated the code to reflect that, and once Claire's changes go in, we can move that candidate filtering stuff to the rules.

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: re: accessing global information in the skip rules, I went on a journey on this one and my conclusion is that it is useful to access this information, but that it's not _actually_ global. In an example (C depends on A and B, all partitioned, everything is fully materialized):

Tick 1: lots of partitions of A are updated, resulting in lots of materialize_on_parent_updated rules firing. However, all of the partitions of B are currently outdated (or basically any other skip reason causes none of the partitions of C to fire).

Tick 2: we remember all of those materialize reasons from the previous tick, and start feeding them through our skip rules. One of those skip rules is (e.g.) skip_on_parent_missing. If we only look at the previous evaluation data for that specific rule, we won't find anything, as none of the parent partitions are actually missing. However, we know that all of those materialize reasons existed the previous tick, and therefore the skip_on_parent_missing rule must have evaluated those partitions, and so the fact that there's no evaluation data implies that it evaluated to false on the previous tick.

In the updated code, we're really just looking at the set of asset partitions which had at least one materialize reason on the previous tick, as all of those would have ended up being evaluated by all skip rules on the previous tick.

I'm saying that this is not *really* global / cross-rule information, as the thing we're using it for is specific to our rule (which asset partitions were evaluated by this rule last tick), it just so happens that the way we calculate that value is by looking at other rules, and that this value is constant across all skip rules.

[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: > If we only look at the previous evaluation data for that specific rule, we won't find anything, as none of the parent partitions are actually missing.

Is my understanding correct that the issue here is that the data stored inside `latest_evaluation_by_asset_key` doesn't allow us to distinguish between these two cases?
- Situations where we evaluated the skip rule on asset partition X, and the partition wasn't skipped.
- Situations where we never evaluated the skip rule on asset partition X.

[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: That's right yep, although I'd phrase it more specifically along the lines of "there is no explicit metadata written when a skip rules decides not to skip an asset partition, and therefore we have to infer the fact that it decided not to skip a partition by inferring the set of partitions it must have evaluated"

[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: Got it.

>  I went on a journey on this one and my conclusion is that it is useful to access this information, but that it's not actually global.

I think future readers will find this difficult to decipher, and it seems very easy for global logic to sneak in.

If there's no straightforward / performant way to track this on a cursor, could we factor this out into a helper method? E.g.

```python
class RuleEvaluationContext(...):
    def was_evaluated_on_last_tick(partition_key: Optional[str]) -> bool:
        """
        <Explanation of the "However, we know that all of those materialize reasons existed the previous tick, and therefore the skip_on_parent_missing rule must have evaluated those partitions, and so the fact that there's no evaluation data implies that it evaluated to false on the previous tick" reasoning> 
        """
        ...
```

Then, in the future, if we can find a way to replace this with an implementation that relies on fewer assumptions, we won't need to re-reason about as many implications.

[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: @sryza currently this is handled by `get_candidates_not_evaluated_on_previous_tick`, which I think is in the shape you're looking for, but should be renamed something more specific to skip rules

[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: Ah ok oops yes that basically solves it. I think we should just find a way to make it clear (a couple words in the docstring would be fine) that this is referring to being evaluated in this specific rule.

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py [hunk 8]
```diff
@@ -350,16 +351,15 @@ def evaluate_asset(
                 to_materialize.update(asset_partitions)
             self._logger.debug("Done evaluating materialize rule")
 
-        # These should be conditions, but aren't currently, so we just manually strip out things
-        # from our materialization set
+        # TODO: make this a skip rule
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:354]: Is this comment change relevant to this PR?

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py [hunk 9]
```diff
@@ -150,17 +150,17 @@ def with_updates(
                 "Latest storage ID should be >= previous latest storage ID",
             )
 
-        result_latest_evaluation_by_asset_key = {**self.latest_evaluation_by_asset_key}
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py:153]: This should be part of the underlying PR, right?

Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py:153]: ah yeah I'll port it over

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 10]
```diff
@@ -397,15 +514,20 @@ def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationRe
...
                     )
                 ].add(asset_partition)
-        if conditions:
-            return [(k, v) for k, v in conditions.items()]
-        return []
+
+        return self.add_evaluation_data_from_previous_tick(
+            context,
+            asset_partitions_by_evaluation_data,
+            should_use_past_data_fn=lambda ap: not context.materialized_requested_or_discarded_since_previous_tick(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:527]: What if there's an in-progress run targeting the asset materialization which wasn't requested by auto-materialize? Does this PR change the handling in that case? Just checking.

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 11]
```diff
@@ -215,6 +318,35 @@ def description(self) -> str:
...
 
+    def add_evaluation_data_from_previous_tick(
+        self,
+        context: RuleEvaluationContext,
+        asset_partitions_by_evaluation_data: Mapping[
+            Optional[AutoMaterializeRuleEvaluationData], Set[AssetKeyPartitionKey]
+        ],
+        should_use_past_data_fn: Callable[[AssetKeyPartitionKey], bool],
+    ) -> "RuleEvaluationResults":
+        """Combines a given set of evaluation data with evaluation data from the previous tick.
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:329]: Maybe add something like "The returned value will include the union of asset partitions that are in `asset_partitions_by_evaluation_data` and asset partitions that the previous tick decided to materialize, minus asset partitions for which `should_use_past_data_fn` evaluates to `True`"

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 12]
```diff
@@ -660,19 +784,39 @@ def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationRe
...
-            ]
+        return self.add_evaluation_data_from_previous_tick(
+            context,
+            asset_partitions_by_evaluation_data,
+            should_use_past_data_fn=lambda ap: ap not in candidates_to_evaluate,
+        )
+
+
+@whitelist_for_serdes
+class DiscardOnMaxMaterializationsExceededRule(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:799]: Is this a relic from a different PR?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 13]
```diff
@@ -751,6 +895,117 @@ def from_rule_evaluation_results(
...
+        """This function returns if a stored record is equivalent to this one. To do so, we can't
+        just use regular namedtuple equality, as the serialized partition subsets will be
+        potentially have different string values.
+        """
+        if stored_evaluation is None:
+            return False
+        sorted_results = sorted(self.partition_subsets_by_condition)
+        sorted_stored_results = sorted(stored_evaluation.partition_subsets_by_condition)
+        return (
+            set(self.rule_snapshots or []) == set(stored_evaluation.rule_snapshots or [])
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:991]: If we add fields to `AutoMaterializeAssetEvaluation`, it seems like we it will be easy to forget to handle them here. Anything we can do to mitigate that?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 14]
```diff
@@ -751,6 +895,117 @@ def from_rule_evaluation_results(
...
+        return self._get_asset_partitions_with_decision_type(
+            AutoMaterializeDecisionType.MATERIALIZE, asset_graph
+        )
+
+    def equivalent_to_stored_evaluation(
+        self, stored_evaluation: Optional["AutoMaterializeAssetEvaluation"], asset_graph: AssetGraph
+    ) -> bool:
+        """This function returns if a stored record is equivalent to this one. To do so, we can't
+        just use regular namedtuple equality, as the serialized partition subsets will be
+        potentially have different string values.
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:984]: What's a situation where they would have different string values? Is it due to the undeterministic ordering of partitions? I wonder if we can just make this more deterministic.

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 15]
```diff
@@ -751,6 +895,117 @@ def from_rule_evaluation_results(
                 rule_snapshots=auto_materialize_policy.rule_snapshots,
             )
 
+    def _deserialize_result(
+        self,
+        serialized_result: Tuple[
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:900]: This name is confusing to me, because I would expect an object named `serialized_x` to be a `str` or byte buffer.

Would it make sense to have separate `rule_evaluation` and `serialized_partitions_subset` arguments?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py [hunk 16]
```diff
@@ -751,6 +895,117 @@ def from_rule_evaluation_results(
                 rule_snapshots=auto_materialize_policy.rule_snapshots,
             )
 
+    def _deserialize_result(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:898]: Maybe `_deserialize_rule_evaluation_result`?

**Review by sryza — COMMENTED** [review]
Looks very close!

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 12 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:799]: yeah this class ended up underneath the most recent skip rule, this is the correct place for it but I'll just move it back to make the diff nicer (I was planning on doing some other cleanup here anyway so I can do that in the next PR)

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 14 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:984]: Yep this is purely due to undeterministic string orderings. I can make them deterministic in a followup and remove this line there

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 13 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:991]: hm we could in theory just keep a set of "excluded fields" and "fields we need to check for equality in a weird way", then just do the normal equality check for the remainder of the fields, but for what it's worth, I think the `asset_key` / `partition_subsets_by_condition` / `rule_snapshots` essentially fully determine the equivalence of these two objects (the num_x are basically derived properties from those), so I think that might end up being a bit over-engineered for now

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 10 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:530]: No change in behavior here -- `requested` only filters out runs requested by the asset daemon, which mirrors the current behavior.

Definitely possible that this is not the ideal behavior, but at least we're not changing it

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 13 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: Once we have deterministic serialization for partition subsets, do we need this? Or can we just compare the strings?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 10 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py:530]: Cool

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py (hunk 13 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: ah missed this question -- the only remaining bit is that when this gets deserialized, partition_subsets_by_condition gets deserialized as a list of lists, rather than a list of tuples, but that could be fixed by overriding the __new__ on this, so yeah this could be totally removed I believe

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py]: nice

**Review by sryza — APPROVED** [review]
Ok bombs away!

**Comment by sryza** [comment]
!!z
